snap/validate.go

```diff
@@ -71,6 +73,92 @@ func ValidateAlias(alias string) error {
 	return nil
 }
 
+// ValidateAppSocketName checks if a string ca be used as a name for a socket
+// (for socket activation).
+func ValidateAppSocketName(name string) error {
+	valid := validSnapName.MatchString(name)
+	if !valid {
+		return fmt.Errorf("invalid socket name: %q", name)
+	}
+	return nil
+}
+
+// ValidateAppSocketListenAddress checks that the value of socket addresses.
+func ValidateAppSocketListenAddress(socket *SocketInfo, fieldName string, address string) error {
+	if address == "" {
+		return fmt.Errorf("socket %q must define %q", socket.Name, fieldName)
+	}
+
+	switch address[0] {
+	case '/', '$':
+		return validateAppSocketListenAddressPath(socket, fieldName, address)
+	case '@':
+		return validateAppSocketListenAddressAbstractSocket(socket, fieldName, address)
+	default:
+		return validateAppSocketListenAddressNetAddress(socket, fieldName, address)
+	}
+}
+
+func validateAppSocketListenAddressPath(socket *SocketInfo, fieldName string, path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf(
+			`socket %q has invalid %q: paths must not include "." or ".."`, socket.Name, fieldName)
+	}
+
+	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/")) {
```

Comment:
also: any reason not to support `/run/snap.${SNAP}.foo`?

Reply:
I think allowing relative path would be a bit confusing, it's not explicit what's the base path in that case ( `$SNAP_DATA`  vs `$SNAP_COMMON` or another dir)

Reply:
Also, the point here is to require to have the path start with `$SNAP_DATA` or `$SNAP_COMMON`,  so that there are no hardcoded paths (like `/var/snap/<mysnap>/common`).
I can add `${SNAP_DATA}`  and `${SNAP_COMMON}` to the list of allowed prefixes, but using `os.Expand` makes the check more complicated and the error message possibly confusing